### PR TITLE
ENH: Better picks

### PR DESCRIPTION
**Pull request by dengemann**
Yes, let's be pickier!

closes https://github.com/autoreject/autoreject/issues/33

### CODE REVIEW

File: autoreject/ransac.py [hunk 1]
```diff
@@ -70,6 +75,9 @@ def __init__(self, n_resample=50, min_channels=0.25, min_corr=0.75,
             Number of parallel jobs.
         random_state : None | int
             To seed or not the random number generator.
+        picks : ndarray, shape(n_channels) | None
+            The channels to be considered for autoreject. If None, defaults
+            to data channels {'meg', 'eeg'}.
```
Comment (jasmainak) [inline autoreject/ransac.py]: cannot be both

File: autoreject/ransac.py [hunk 2]
```diff
@@ -52,7 +56,8 @@ class Ransac(object):
 
     def __init__(self, n_resample=50, min_channels=0.25, min_corr=0.75,
                  unbroken_time=0.4, ch_type='eeg', n_jobs=1,
-                 random_state=435656, verbose='progressbar'):
```
Comment (jasmainak) [inline autoreject/ransac.py]: ch_type is not documented

File: autoreject/tests/test_autoreject.py [hunk 3]
```diff
@@ -33,7 +33,7 @@ def test_autoreject():
 
     include = [u'EEG %03d' % i for i in range(1, 45, 3)]
     picks = mne.pick_types(raw.info, meg=False, eeg=False, stim=False,
-                           eog=True, include=include, exclude=[])
+                           eog=False, include=include, exclude=[])
```
Comment (jasmainak) [inline autoreject/tests/test_autoreject.py]: diff not needed

File: autoreject/tests/test_autoreject.py [hunk 4]
```diff
@@ -67,22 +67,22 @@ def test_autoreject():
...
 
-    ar = LocalAutoReject()
+    ar = LocalAutoReject(picks=picks)
     assert_raises(NotImplementedError, validation_curve, ar, epochs, None,
                   param_name, param_range)
 
-    ar = LocalAutoRejectCV(cv=3)
-    assert_raises(ValueError, ar.fit, X)
+    ar = LocalAutoRejectCV(cv=3, picks=picks)
+    assert_raises(AttributeError, ar.fit, X)
```
Comment (jasmainak) [inline autoreject/tests/test_autoreject.py]: put back the valuerror?

File: autoreject/tests/test_autoreject.py [hunk 5]
```diff
@@ -67,22 +67,22 @@ def test_autoreject():
...
                   param_name, param_range)
 
-    ar = LocalAutoReject()
+    ar = LocalAutoReject(picks=picks)
     assert_raises(NotImplementedError, validation_curve, ar, epochs, None,
                   param_name, param_range)
 
-    ar = LocalAutoRejectCV(cv=3)
-    assert_raises(ValueError, ar.fit, X)
+    ar = LocalAutoRejectCV(cv=3, picks=picks)
```
Comment (jasmainak) [inline autoreject/tests/test_autoreject.py]: make sure channels are consistent between fit and transform

File: autoreject/tests/test_autoreject.py [hunk 6]
```diff
@@ -67,22 +67,22 @@ def test_autoreject():
...
+    assert_raises(AttributeError, ar.fit, X)
     assert_raises(ValueError, ar.transform, X)
     assert_raises(ValueError, ar.transform, epochs)
 
     ar.fit(epochs)
-    assert_true(len(ar.picks) == len(picks) - 1)
```
Comment (jasmainak) [inline autoreject/tests/test_autoreject.py]: test for EOG + EEG?

File: autoreject/tests/test_autoreject.py [hunk 7]
```diff
@@ -94,10 +94,16 @@ def test_autoreject():
...
-        ValueError, ar.transform,
+        IndexError, ar.transform,
         epochs.copy().pick_channels(
             [epochs.ch_names[pp] for pp in picks[:3]]))
 
+    assert_equal(ar.bad_segments.shape[1], len(epochs.ch_names))
+    assert_true(np.any(ar.bad_segments[:, picks]))
+    anti_picks = np.ones(len(epochs.ch_names), dtype=bool)
+    anti_picks[picks] = False
+    assert_true(not np.any(ar.bad_segments[:, anti_picks]))
```
Comment (jasmainak) [inline autoreject/tests/test_autoreject.py]: https://www.youtube.com/watch?v=x-64CaD8GXw

File: autoreject/utils.py [hunk 8]
```diff
@@ -248,17 +248,29 @@ def get_picks_bad_good(info):
...
-    _do_interp_dots(inst, mapping, picks_good, picks_bad_)
+    if not inst_picked:
+        picks_good_ = [pp for pp in picks if pp in picks_good_]
+    assert len(picks_good_) == len(picks_good)
+
+    # XXX all trouble is probably here
+    inst._data[:, picks_bad_orig, :] = (
+        # multiply correctly with interpolation matrix
+        np.einsum('ij,xjy->xiy',  # subpick data
+                  mapping, inst._data[:, picks_good_]))
```
Comment (dengemann) [inline autoreject/utils.py]: @jasmainak can you take a look? I think somehow the remapping between the picked info from where we get the mapping the the unpicked instance is not correct. Tried to leave some comments. Debugging this is difficult as long as we don't  have a regression test / test against ground truth. Currently the symptom is imply that the right bad channel is not repaired.

See here the output from plot_auto_repair.py

screenshot 2017-06-25 22 51 37

However, the code before seems to work as we have clearly a bad channel here:

screenshot 2017-06-25 22 56 15

File: autoreject/autoreject.py [hunk 9]
```diff
@@ -563,9 +566,10 @@ def _interpolate_bad_epochs(self, epochs, ch_type, verbose='progressbar'):
             bad_chs = np.where(bad_chs_mask)[0]
             bad_chs = [ch_name for idx, ch_name in enumerate(ch_names)
                        if idx in bad_chs]
+
             epoch = epochs[epoch_idx]
             epoch.info['bads'] = bad_chs
-            interpolate_bads(epoch, reset_bads=True)
+            interpolate_bads(epoch, picks=self.picks, reset_bads=True)
```
Comment (jasmainak) [inline autoreject/autoreject.py]: test against MNE-python interpolation

File: autoreject/utils.py [hunk 10]
```diff
@@ -139,49 +140,139 @@ def fetch_file(url, file_name, resume=True, timeout=10.):
...
+    ch_names_a = [picked_info['ch_names'][pp] for pp in picks_bad]
+    ch_names_b = [inst.info['ch_names'][pp] for pp in picks_bad_orig]
+    assert ch_names_a == ch_names_b
+    if not inst_picked:
+        picks_good_ = [pp for pp in picks if pp in picks_good_]
+    assert len(picks_good_) == len(picks_good)
+    ch_names_a = [picked_info['ch_names'][pp] for pp in picks_good]
+    ch_names_b = [inst.info['ch_names'][pp] for pp in picks_good_]
+    assert ch_names_a == ch_names_b
+    # XXX all trouble is probably here
```
Comment (jasmainak) [inline autoreject/utils.py]: you can remove this comment now

File: examples/plot_channel_thresholds.py [hunk 11]
```diff
@@ -71,7 +69,8 @@
 plt.tick_params(axis='x', which='both', bottom='off', top='off')
 plt.tick_params(axis='y', which='both', left='off', right='off')
 
-plt.hist(scaling * np.array(threshes.values()), 30, color='g', alpha=0.4)
+plt.hist(scaling * np.array(list(threshes.values())), 30,
```
Comment (jasmainak) [inline examples/plot_channel_thresholds.py]: why do you need `list()`? Is it a python 2 vs python 3 thing?

**Comment by dengemann** [comment]
python 3 dude! dict_values can't go in array.

File: autoreject/ransac.py [hunk 12]
```diff
@@ -148,8 +152,9 @@ def _get_mappings(self, inst):
                                  for name in ch_subsets[idx]])
             mapping = np.zeros((n_channels, n_channels))
             if self.ch_type == 'meg':
-                mapping[:, pick_from] = _fast_map_meg_channels(inst, pick_from,
-                                                               pick_to)
+                # XXX Mainak here we need to similar things as in AR
+                mapping[:, pick_from] = _fast_map_meg_channels(
+                    inst.info.copy(), pick_from, pick_to)
```
Comment (jasmainak) [inline autoreject/ransac.py]: merci! sorry I couldn't make a lot of progress. Got interrupted many times. I'll try to continue after dinner if you don't beat me to it

**Comment by dengemann** [comment]
@jasmainak convince yourself that we're done here :) ransac fixed with latest commit! Wooot! Green lights for pickigng. And the fun can begin.

File: autoreject/ransac.py [hunk 13]
```diff
@@ -137,38 +152,41 @@ def _get_mappings(self, inst):
                                  for name in ch_subsets[idx]])
             mapping = np.zeros((n_channels, n_channels))
             if self.ch_type == 'meg':
-                mapping[:, pick_from] = _fast_map_meg_channels(inst, pick_from,
-                                                               pick_to)
+                # XXX Mainak here we need to similar things as in AR
```
Comment (jasmainak) [inline autoreject/ransac.py]: comment not needed any more

**Comment by dengemann** [comment]
@agramfort @fraimondo can you give us a hand here and help us decide whether this PR is equivalent with master and does preserve the functionality we expose so far. Please run http://autoreject.github.io/auto_examples/plot_visualize_bad_epochs.html#sphx-glr-auto-examples-plot-visualize-bad-epochs-py on master and on this branch and report differences to the doc build on the website.

**Comment by dengemann** [comment]
The background is that seed / randomization issues potentially make results different across computers.

**Comment by jasmainak** [comment]
@dengemann I think there is one thing in PR https://github.com/autoreject/autoreject/pull/45/ that we have ignored so far. When we count the number of bad segments, we simply sum without picking. You need to think carefully what happens when we incrementally update `ar.bad_segments`

**Comment by fraimondo** [comment]
On Master I get 3.1% of bad epochs, on branch, 6.2%.
Master
master

Branch
branch

**Comment by dengemann** [comment]
@jasmainak it seems your master branch fix did the magic. Now solutions seem stable between and within branches.

**Comment by jasmainak** [comment]
@dengemann you have a flake8 error

File: examples/plot_channel_thresholds.py [hunk 14]
```diff
@@ -53,7 +50,8 @@
 import numpy as np  # noqa
 from autoreject import compute_thresholds  # noqa
 
-threshes = compute_thresholds(epochs, method='random_search',
+threshes = compute_thresholds(epochs, picks=picks,
+                              method='bayesian_optimization',
```
Comment (jasmainak) [inline examples/plot_channel_thresholds.py]: can you revert this to `'random_search'` ?

File: autoreject/utils.py [hunk 15]
```diff
@@ -220,7 +282,10 @@ def _compute_dots(info, mode='fast'):
     my_origin = _check_origin((0., 0., 0.04), info_from)
     int_rad, noise, lut_fun, n_fact = _setup_dots(mode, coils_from, 'meg')
 
-    self_dots, cross_dots = _compute_fast_dots(info, mode=mode)
+    # this function needsa clean input. It hates the presence of other
```
Comment (jasmainak) [inline autoreject/utils.py]: not sure I understand this comment. Let's remove it?

File: autoreject/utils.py [hunk 16]
```diff
@@ -2,12 +2,40 @@
 
 # Authors: Mainak Jas <mainak.jas@telecom-paristech.fr>
 
+import numpy as np
 import mne
+from mne.utils import logger
```
Comment (jasmainak) [inline autoreject/utils.py]: @dengemann we can't use logger from another package. There is no control. Might as well use a print statement

File: autoreject/tests/test_autoreject.py [hunk 17]
```diff
@@ -25,64 +25,106 @@
...
+
+        ar.fit(epochs)
+        # Now we test that the .bad_segments has the shape
+        # of n_trials, n_sensors, such that n_sensors is the
+        # the full number sensors, before picking. We, hence,
+        # expect nothing to be rejected outside of our picks
+        # but rejections can occur inside our picks.
+        assert_equal(ar.bad_segments.shape[1], len(epochs.ch_names))
+        assert_true(np.any(ar.bad_segments[:, picks]))
+        anti_picks = np.ones(len(epochs.ch_names), dtype=bool)
```
Comment (jasmainak) [inline autoreject/tests/test_autoreject.py]: same as what we call `non_picks` elsewhere?

File: autoreject/tests/test_autoreject.py [hunk 18]
```diff
@@ -25,64 +25,106 @@
...
+
+        picks = mne.pick_types(
+            epochs.info, meg=meg, eeg=eeg, stim=False, eog=False,
+            include=include_, exclude=[])
+
+        X = epochs.get_data()
+        n_epochs, n_channels, n_times = X.shape
+        X = X.reshape(n_epochs, -1)
+
+        if ii == 0:
```
Comment (jasmainak) [inline autoreject/tests/test_autoreject.py]: this should be outside the loop? Doesn't seem to depend on `picks` ...

**Comment by jasmainak** [comment]
Thanks @dengemann PR looks good. Last couple of nitpicks and let's move on to the PR with channel types. We are almost there !

**Comment by dengemann** [comment]
Phew. It's been a tough piece of git. Thanks @massich @agramfort @jasmainak for this nice time work.

**Comment by agramfort** [comment]
Great !

**Comment by jasmainak** [comment]
Awesome, now we need to update an example to show the multi-channel support :)

**Pull request merged**